nav2_assisted_teleop/src/nav2_assisted_teleop.cpp

```diff
@@ -0,0 +1,311 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_assisted_teleop/nav2_assisted_teleop.hpp"
+#include "tf2_ros/create_timer_ros.h"
+#include "nav2_util/node_utils.hpp"
+
+#include <thread>
+
+namespace nav2_assisted_teleop
+{
+
+  AssistedTeleop::AssistedTeleop(rclcpp::Logger& logger) : logger_(logger)
+  {}
+
+  void AssistedTeleop::on_configure(std::shared_ptr<nav2_util::LifecycleNode> node,
+                    std::shared_ptr<tf2_ros::Buffer> tf,
+                    std::shared_ptr<nav2_costmap_2d::CostmapTopicCollisionChecker> collision_checker)
+  {
+    // set up parameters
+    node->declare_parameter("transform_tolerance", rclcpp::ParameterValue(0.1));
+
+    nav2_util::declare_parameter_if_not_declared(
+    node, "projection_time", rclcpp::ParameterValue(1.0));
+
+    node->get_parameter("global_frame", global_frame_);
+    node->get_parameter("robot_base_frame", robot_base_frame_);
+
+    // Note: empty frames don't cause robot_utils to error
+    // nav2_util::declare_parameter_if_not_declared(
+    //   node,
+    //   "global_frame", rclcpp::ParameterValue(std::string("global_frame")));
+
+    // nav2_util::declare_parameter_if_not_declared(
+    //   node,
+    //   "robot_base_frame", rclcpp::ParameterValue(std::string("robot_base_frame")));
+
+    tf_ = tf;
+    collision_checker_ = collision_checker;
+  }
+
+  geometry_msgs::msg::Twist AssistedTeleop::computeVelocity(geometry_msgs::msg::Twist& twist)
+  {
+    geometry_msgs::msg::PoseStamped current_pose;
+    if (!nav2_util::getCurrentPose(
+        current_pose, *tf_, global_frame_, robot_base_frame_,
+        transform_tolerance_))
+    {
+      RCLCPP_ERROR(logger_, "Failed to get current pose, setting velocity to zero");
+      return geometry_msgs::msg::Twist();
+    }
+
+    geometry_msgs::msg::Pose2D current_pose_2d;
+    current_pose_2d.x = current_pose.pose.position.x;
+    current_pose_2d.y = current_pose.pose.position.y;
+    current_pose_2d.theta = 0.0; //tf2::getYaw(current_pose.pose.orientation);
+
+    const double linear_vel = std::fabs(std::hypot(twist.linear.x, twist.linear.y));
+    const double dt = 0.05 / linear_vel;
+
+    geometry_msgs::msg::Pose2D projected_pose = current_pose_2d;
+    auto adjusted_twist = twist;
+    for(double current_time = 0; current_time < projection_time_; current_time+=dt)
+    {
+      projected_pose = projectPose(projected_pose, twist, dt);
+
+      if(!collision_checker_->isCollisionFree(projected_pose))
+      {
+        adjusted_twist.linear.x *= current_time / projection_time_;
```

Comment:
Is `*=` the right thing? Recursively, this would be 0/max, then 0.05/max, then 0.1/max, etc. so each time its not just projecting by the current segment, but all prior as well.

Reply:
I was missing the break:).

Reply:
Not sure I understand the comment, but I think we're on the same page :smile: